pfcpiface/aether.go

```diff
@@ -0,0 +1,849 @@
+			}
+		}
+		// No collision found, this ogate is free.
+		return i, nil
+	}
+
+	return 0, ErrNotFound("free ogate")
+}
+
+func (a *aether) addIPLookupRule(ctx context.Context, dst *net.IPNet, nhopMAC net.HardwareAddr) (err error) {
+	ones, zeros := dst.Mask.Size()
+	if ones+zeros == 0 {
+		return ErrInvalidArgumentWithReason("addIPLookupRule", dst, "not a CIDR mask")
+	}
+
+	// Create next hop dst MAC packet data update module and add to cache, if needed.
+	n, exists := a.neighborCache[dst.String()]
+	log.Tracef("neighbor cache for %v exists: %v, %+v", dst, exists, n)
+	if !exists {
+		n.nhopMAC = nhopMAC
+		n.updateModuleName, err = a.createDstMacUpdateModule(ctx, dst, nhopMAC)
+		if err != nil {
+			return err
+		}
+
+		n.ogate, err = a.findNextFreeOgate(ctx, datapathIPLookModule)
+		if err != nil {
+			return err
+		}
+
+		// Place module in between the IP lookup and merge modules.
+		if err = a.linkBessModules(ctx, datapathIPLookModule, n.ogate, n.updateModuleName, 0); err != nil {
+			return err
+		}
+		if err = a.linkBessModules(ctx, n.updateModuleName, 0, datapathIfaceName+"Merge", 0); err != nil {
+			return err
+		}
+
+		// Insert lookup rule.
+		msg := &pb.IPLookupCommandAddArg{
+			Prefix:    dst.IP.Mask(dst.Mask).String(), // Need to clear lower bits.
+			PrefixLen: uint64(ones),
+			Gate:      n.ogate,
+		}
+		if err = a.processIPLookup(ctx, msg, moduleMethodAdd); err != nil {
+			return err
+		}
+	}
+	n.routeCount++
+	a.neighborCache[dst.String()] = n
+
+	return nil
+}
+
+func (a *aether) deleteIPLookupRule(ctx context.Context, dst *net.IPNet) error {
+	ones, zeros := dst.Mask.Size()
+	if ones+zeros == 0 {
+		return ErrInvalidArgumentWithReason("deleteIPLookupRule", dst, "not a CIDR mask")
+	}
+
+	// Check if neighbor has a module and delete if necessary.
+	n, exists := a.neighborCache[dst.String()]
+	log.Tracef("neighbor cache for %v exists: %v, %+v", dst, exists, n)
+	if exists {
+		n.routeCount--
+		a.neighborCache[dst.String()] = n
+
+		if n.routeCount == 0 {
+			delete(a.neighborCache, dst.String())
+			if err := a.deleteModule(ctx, n.updateModuleName); err != nil {
+				return err
+			}
+			// Delete lookup rule.
+			msg := &pb.IPLookupCommandDeleteArg{
+				Prefix:    dst.IP.String(), // Need to clear lower bits.
+				PrefixLen: uint64(ones),
+			}
+			if err := a.processIPLookup(ctx, msg, moduleMethodDelete); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (a *aether) processIPLookup(ctx context.Context, msg proto.Message, method string) error {
+	switch method {
+	case moduleMethodAdd:
+		fallthrough
+	case moduleMethodDelete:
+		fallthrough
+	case moduleMethodClear:
+	default:
+		return ErrInvalidArgumentWithReason("method", method, "invalid method name")
+	}
+
+	any, err := anypb.New(msg)
+	if err != nil {
+		log.Error("Error marshalling the rule", msg, err)
+		return err
+	}
+
+	// IPLookup module is not thread-safe, need to pause processing.
+	if err := a.pauseBessWorkers(ctx); err != nil {
```

Comment:
Again the comment about deferring the resume operation. It might be even more unsafe in case of IP lookup rules as they seem to be added/removed at runtime and dynamically.